drivers/hmis/app/models/hmis/project_unit_type_mapping.rb

```diff
@@ -17,32 +17,72 @@ class Hmis::ProjectUnitTypeMapping < Hmis::HmisBase
+
+    # Collect unit groups to create
+    unit_groups_to_create = records_needing_new_units.map do |record|
+      {
+        project_id: record.project.id,
+        unit_type_id: record.unit_type.id,
+        name: record.unit_type.description,
+      }
+    end
+
+    # Batch import unit groups. These probably don't already exist,
+    # since we checked that no units exist for this project and unit type.
+    # But, it could exist and be empty, so just in case, ignore on conflict.
+    Hmis::UnitGroup.import!(unit_groups_to_create, on_duplicate_key_ignore: true)
+
+    # Get unit group IDs and put them in a hash keyed by [project_id, unit_type_id] for lookup when creating units
+    unit_groups_by_key = Hmis::UnitGroup.where(project: records_needing_new_units.map(&:project), unit_type: records_needing_new_units.map(&:unit_type)).
+      pluck(:project_id, :unit_type_id, :id).
+      to_h { |project_id, unit_type_id, id| [[project_id, unit_type_id], id] }
+
+    # Collect units to create
+    units_to_create = records_needing_new_units.flat_map do |record|
       project = record.project
-      next if project.nil? # could happen if project was deleted but ProjectUnitTypeMapping wasn't properly cleaned up
+      unit_type = record.unit_type
 
-      # If this ProjectID is already mapped to this UnitTypeID in our system, and it is marked as Active=Y, do nothing.
-      key = [project.id, unit_type.id]
-      next if unit_counts_by_project_and_unit_type_id[key]
+      unit_group_id = unit_groups_by_key[[project.id, unit_type.id]]
+      raise "Failed to create unit group for project  #{project.id}, unit_type #{unit_type.id}" unless unit_group_id
 
-      # If this ProjectID is not already mapped to this UnitTypeID, add it and add the number of units specified in the UnitCapacity column.
+      # Create the number of units specified in the UnitCapacity column
       record.unit_capacity.to_i.times.map do
         {
           project_id: project.id,
           unit_type_id: unit_type.id,
+          hmis_unit_group_id: unit_group_id,
           unit_size: unit_type.unit_size,
           user_id: user.id,
         }
       end
     end
-    Hmis::Unit.import!(unit_attrs.compact, validate: false)
 
-    # If this ProjectID is already mapped to this UnitTypeID in our system, but it is marked as Active=N, then remove the mapping. At this point also remove any Units for this unit type. Raise if any of those Units are occupied
+    # Batch import units
+    Hmis::Unit.import!(units_to_create, validate: false)
+  end
+
+  def self.destroy_inactive_units(scope, today: Date.current)
+    # If this ProjectID is already mapped to this UnitTypeID in our system, but it is marked as Active=N, then remove the mapping.
+    # At this point also remove any Units for this unit type. Raise if any of those Units are occupied
     scope.filter(&:inactive?).each do |record|
       existing_units = record.project.units.where(unit_type: record.unit_type)
```

Comment:
Maybe add a TODO here for `#8157`? Since this will need to be changed to look up using the unit group-unit-type relationship, ie something like `record.project.unit_groups.where(unit_type: record.unit_type).each(&:destroy!)`